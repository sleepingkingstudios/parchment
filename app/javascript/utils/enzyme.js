import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
