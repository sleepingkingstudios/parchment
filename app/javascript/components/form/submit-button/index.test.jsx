import React from 'react';
import { shallow } from 'enzyme';

import FormSubmitButton from './index';

describe('<FormSubmitButton />', () => {
  const value = 'Property Value';
  const data = { propertyName: value };
  const onSubmitAction = jest.fn(() => ({ payload: { ok: true } }));
  const form = {
    data,
    onSubmitAction,
  };
  const defaultProps = { form };

  it('should render the button', () => {
    const rendered = shallow(<FormSubmitButton {...defaultProps} />);

    expect(rendered).toHaveDisplayName('Button');
    expect(rendered).toHaveProp({ children: 'Submit Form' });
  });

  it('should set the onClick event handler', () => {
    const rendered = shallow(<FormSubmitButton {...defaultProps} />);
    const { onClick } = rendered.props();
    const preventDefault = jest.fn();
    const event = { preventDefault };

    expect(typeof onClick).toEqual('function');

    const action = onClick(event);

    expect(preventDefault).toHaveBeenCalled();
    expect(onSubmitAction).toHaveBeenCalledWith();
    expect(action).toEqual({ payload: { ok: true } });
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<FormSubmitButton {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });

  describe('with actionName: value', () => {
    const actionName = 'Execute';

    it('should render the button', () => {
      const rendered = shallow(<FormSubmitButton {...defaultProps} actionName={actionName} />);

      expect(rendered).toHaveDisplayName('Button');
      expect(rendered).toHaveProp({ children: 'Execute Form' });
    });
  });

  describe('with children: value', () => {
    const children = 'Execute Operation';

    it('should render the button', () => {
      const rendered = shallow(<FormSubmitButton {...defaultProps}>{children}</FormSubmitButton>);

      expect(rendered).toHaveDisplayName('Button');
      expect(rendered).toHaveProp({ children });
    });
  });

  describe('with resourceName: value', () => {
    const resourceName = 'Operation';

    it('should render the button', () => {
      const rendered = shallow(<FormSubmitButton {...defaultProps} resourceName={resourceName} />);

      expect(rendered).toHaveDisplayName('Button');
      expect(rendered).toHaveProp({ children: 'Submit Operation' });
    });
  });
});
