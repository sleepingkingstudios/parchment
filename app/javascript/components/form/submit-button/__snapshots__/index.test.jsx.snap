// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSubmitButton /> should match the snapshot 1`] = `
<Button
  block={true}
  buttonSize="md"
  buttonStyle="primary"
  className="form-submit"
  disabled={false}
  link={false}
  onClick={[Function]}
  outline={true}
>
  Submit Form
</Button>
`;
